gobblin-runtime/src/main/java/org/apache/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -392,14 +395,30 @@ private void addJars(Path jarFileDir, String jarFileList, Configuration conf)
     for (String jarFile : SPLITTER.split(jarFileList)) {
       Path srcJarFile = new Path(jarFile);
       FileStatus[] fileStatusList = lfs.globStatus(srcJarFile);
+
+      jarFileLoop:
       for (FileStatus status : fileStatusList) {
-        // SNAPSHOT jars should not be shared, as different jobs may be using different versions of it
-        Path baseDir = status.getPath().getName().contains("SNAPSHOT") ? this.unsharedJarsDir : jarFileDir;
-        // DistributedCache requires absolute path, so we need to use makeQualified.
-        Path destJarFile = new Path(this.fs.makeQualified(baseDir), status.getPath().getName());
-        if (!this.fs.exists(destJarFile)) {
-          // Copy the jar file from local file system to HDFS
-          this.fs.copyFromLocalFile(status.getPath(), destJarFile);
+        // For each FileStatus there are chances it could fail in copying at the first attempt, due to file-existence
+        // or file-copy is ongoing by other job instance since all Gobblin jobs share the same jar file directory.
+        // the retryCount is to avoid cases (if any) where retry is going too far and causes job hanging.
+        int retryCount = 0 ;
+        Path destJarFile = calculateDestJarFile(status, jarFileDir);
+          // Adding destJarFile into HDFS until it exists and the size of file on targetPath matches the one on local path.
+        while(!this.fs.exists(destJarFile) || fs.listStatus(destJarFile).length != status.getLen()) {
+          try {
+            // Set the first parameter as false for not deleting sourceFile
+            // Set the second parameter as false for not overwriting existing file on the target, by default it is true.
+            // If the file is preExisted but overwrite flag set to false, then an IOException if thrown.
+            this.fs.copyFromLocalFile(false, false, status.getPath(), destJarFile);
+          } catch (IOException ioe) {
+            LOG.warn("Path:" + destJarFile + " is not copied successfully. Will require retry.");
+            retryCount += 1 ;
+            if (retryCount >= MAXIMUM_JAR_COPY_RETRY_TIMES) {
+              LOG.error("The jar file:" + destJarFile + " has been successfully copied into hdfs");
+              // If retry reaches upper limit, skip copying this file. 
+              continue jarFileLoop;
```

Comment:
Should break here to avoid a need for the label.

Reply:
If break, line 423 and 424 will still be executed. Is that OK in this case ?